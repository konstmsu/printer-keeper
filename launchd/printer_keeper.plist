<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>Label</key>
    <string>printer_keeper</string>
    <key>ProgramArguments</key>
    <array>
      <string>/Users/ko/.cargo/bin/uv</string>
      <string>run</string>
      <string>-m</string>
      <string>printer_keeper</string>
    </array>
    <key>StartCalendarInterval</key>
    <array>
      <dict>
        <key>Hour</key>
        <integer>7</integer>
        <key>Minute</key>
        <integer>0</integer>
      </dict>
    </array>
    <key>EnvironmentVariables</key>
    <dict>
      <key>SEND_TO_PRINTER</key>
      <string>1</string>
    </dict>
    <key>WorkingDirectory</key>
    <string>/Users/ko/projects/printer-keeper</string>
    <key>StandardOutPath</key>
    <string>stdout.log</string>
    <key>StandardErrorPath</key>
    <string>stderr.log</string>
  </dict>
</plist>