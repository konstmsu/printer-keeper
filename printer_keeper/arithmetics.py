from dataclasses import dataclass
from random import Random
from typing import List


@dataclass(frozen=True)
class ArithmeticProblem:
    text: str


class ArithmeticProblemGenerator:
    def __init__(self, random_seed=None):
        self.rnd = Random(random_seed)

    def generate(self):
        yield self._sum()
        yield self._sum()
        yield self._difference()
        yield self._difference()
        yield self._multiplication()
        yield self._multiplication()
        yield self._division()
        yield self._division()

    def _sum(self):
        a = self.rnd.randint(8, 20)
        b = self.rnd.randint(5, 40)
        return ArithmeticProblem(f"{a} + {b} = ")

    def _difference(self):
        while True:
            a = self.rnd.randint(8, 40)
            b = self.rnd.randint(5, 30)
            if a - b >= 1:
                return ArithmeticProblem(f"{a} - {b} = ")

    def _multiplication(self):
        while True:
            a = self.rnd.randint(3, 9)
            b = self.rnd.randint(4, 9)
            return ArithmeticProblem(f"{a} * {b} = ")

    def _division(self):
        while True:
            a = self.rnd.randint(4, 9)
            b = self.rnd.randint(3, 9)
            return ArithmeticProblem(f"{a * b} / {b} = ")
