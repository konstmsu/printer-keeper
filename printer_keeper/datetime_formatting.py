from datetime import date, datetime
from typing import Literal


def format_date(d: date, *, include_year: bool) -> str:
    def year():
        if d.year == 2022:
            return "две тысячи двадцать второго"
        if d.year == 2023:
            return "две тысячи двадцать третьего"
        if d.year == 2024:
            return "две тысячи двадцать четвёртого"
        if d.year == 2025:
            return "две тысячи двадцать пятого"
        if d.year == 2026:
            return "две тысячи двадцать шестого"
        # TODO make generic
        raise NotImplementedError

    months = [
        "января",
        "февраля",
        "марта",
        "апреля",
        "мая",
        "июня",
        "июля",
        "августа",
        "сентября",
        "октября",
        "ноября",
        "декабря",
    ]
    days = [
        "первое",
        "второе",
        "третье",
        "четвёртое",
        "пятое",
        "шестое",
        "седьмое",
        "восьмое",
        "девятое",
        "десятое",
        "одиннадцатое",
        "двенадцатое",
        "тринадцатое",
        "четырнадцатое",
        "пятнадцатое",
        "шестнадцатое",
        "семнадцатое",
        "восемнадцатое",
        "девятнадцатое",
        "двадцатое",
        "двадцать первое",
        "двадцать второе",
        "двадцать третье",
        "двадцать четвёртое",
        "двадцать пятое",
        "двадцать шестое",
        "двадцать седьмое",
        "двадцать восьмое",
        "двадцать девятое",
        "тридцатое",
        "тридцать первое",
    ]

    parts = [days[d.day - 1], months[d.month - 1]]

    if include_year:
        parts.append(f"{year()} года")

    return " ".join(parts)


def format_datetime(dt: datetime) -> str:
    simple_numbers = [
        "ноль",
        ("один", "одна"),
        ("два", "две"),
        "три",
        "четыре",
        "пять",
        "шесть",
        "семь",
        "восемь",
        "девять",
        "десять",
        "одиннадцать",
        "двенадцать",
        "тринадцать",
        "четырнадцать",
        "пятнадцать",
        "шестнадцать",
        "семнадцать",
        "восемнадцать",
        "девятнадцать",
    ]
    tens = [
        "двадцать",
        "тридцать",
        "сорок",
        "пятьдесят",
        "шестьдесят",
        "семьдесят",
        "восемьдесят",
        "девяноста",
    ]

    def number(v: int, *, gender: Literal["male", "female"]) -> str:
        gender_index = {"male": 0, "female": 1}[gender]
        if v in [1, 2]:
            return simple_numbers[v][gender_index]
        if 3 <= v < 20:
            return simple_numbers[v]
        if v < 100:
            t = tens[v // 10 - 2]
            if v % 10 == 0:
                return t
            d = number(v % 10, gender=gender)
            return f"{t} {d}"
        return str(v)

    def inflect(word: str, count: int) -> str:
        if 5 <= count % 100 <= 20:
            c = 2
        elif count % 10 == 1:
            c = 0
        elif count % 10 in [2, 3, 4]:
            c = 1
        else:
            c = 2

        words = [("час", "часа", "часов"), ("минута", "минуты", "минут")]
        for forms in words:
            if word == forms[0]:
                return forms[c]
        return word

    parts = [
        format_date(dt.date(), include_year=True) + ",",
        f"{number(dt.hour, gender='male')} {inflect('час', dt.hour)}",
        f"{number(dt.minute, gender='female')} {inflect('минута', dt.minute)}",
    ]
    return " ".join(parts)
