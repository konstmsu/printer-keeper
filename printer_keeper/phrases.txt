Под лежачий камень и вода не течет.
Без труда не выловишь и рыбку из пруда.
Никогда не совершает ошибок только тот, кто ничего не делает.
Делу – время, потехе – час.
Повторенье – мать ученья.
Кукареку!
Ученье и труд всё перетрут.
Ученье — свет, а неученье — тьма.
Умный в гору не пойдёт, умный гору обойдёт.
Яйца курицу не учат.
Чья бы корова мычала, а твоя бы молчала.
Это ещё бабушка надвое сказала.
Это ещё вилами по воде писано.
Это ещё цветочки, а ягодки впереди.