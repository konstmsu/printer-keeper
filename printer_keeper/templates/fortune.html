<html>
  <head>
    <title>Доброе Утро!</title>
  </head>
  <body>
    <p style="font-size: 4em; white-space: pre-wrap"
      >{% for part in parts -%}{{ part|safe }}{% endfor %}</p
    >
  </body>
</html>
