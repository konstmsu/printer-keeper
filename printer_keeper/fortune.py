import logging
import os
import re
import subprocess
import tempfile
from datetime import datetime
from functools import cache
from pathlib import Path
from random import Random
from typing import List

from jinja2 import Environment, PackageLoader, select_autoescape

from .arithmetics import ArithmeticProblemGenerator
from .datetime_formatting import format_date

logger = logging.getLogger("printer_keeper")

env = Environment(
    loader=PackageLoader("printer_keeper"), autoescape=select_autoescape()
)


class MessageGenerator:
    def __init__(self, *, random_seed: int = None):
        self.rnd = Random(random_seed)

    @cache
    def _get_messages(self):
        phrases = Path(__file__).parent / "phrases.txt"
        stripped = [l.strip() for l in phrases.read_text(encoding="utf8").splitlines()]
        return [l for l in stripped if l]

    def generate(self) -> str:
        return self.rnd.choice(self._get_messages())


class MorningFortuneGenerator:
    def __init__(self, *, random_seed: int = None):
        self.rnd = Random(random_seed)

    def generate(self, *, asof: datetime, wisdom: str, problems: List[str]):
        problems_joined = "\n".join(problems)
        return f"""Ура! Доброе утро!
Сегодня у нас {format_date(asof, include_year=False)} {asof.year}.

Пословица дня:
{wisdom}

Чуть-чуть арифметики:
{problems_joined}

Хорошего дня!"""


class MessageHtmlFormatter:
    def __init__(self, *, random_seed: int = None):
        self.rnd = Random(random_seed)

    def _random_color(self):
        # TODO make random colors given contrast is high enough
        return self.rnd.choice(
            [
                "black",
                "red",
                "green",
                "blue",
                "#77450D",
                "#96290D",
                "#A82255",
                "#5C255C",
                "#5642A6",
                "#0C5174",
                "#004D46",
                "#1D7324",
                "#5A701A",
                "#866103",
                "#7A542E",
            ]
        )

    def format(self, message: str) -> str:
        def format_part(m: re.Match):
            word = m.group("word")
            if word:
                return f"<span style='color:{self._random_color()}'>{word}</span>"
            return m[0]

        parts = [
            format_part(m)
            for m in re.finditer("(?P<word>\\w+)|(?P<others>\\W+)", message)
        ]

        template = env.get_template("fortune.html")
        return template.render(parts=parts)


def main():
    html_path = Path(tempfile.mkstemp(prefix="fortune_", suffix=".html")[1])
    pdf_path = Path(tempfile.mkstemp(prefix="fortune_", suffix=".pdf")[1])

    asof = datetime.now()
    message = MessageGenerator().generate()
    arithmetic_problems = ArithmeticProblemGenerator().generate()

    logger.info("Generating fortune asof %s for %s", message, asof)
    fortune = MorningFortuneGenerator().generate(
        wisdom=message, asof=asof, problems=[p.text for p in arithmetic_problems]
    )

    logger.info("Writing HTML to %s", html_path)
    html = MessageHtmlFormatter().format(fortune)
    html_path.write_text(html, encoding="utf8")

    # Convert to PDF. Or is there an easier way to print HTML?
    logger.info("Converting to PDF in %s", pdf_path)
    subprocess.run(
        [
            "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
            "--headless",
            "--disable-gpu",
            f"--print-to-pdf={pdf_path.absolute()}",
            str(html_path.absolute()),
        ],
        check=True,
    )

    # send to printer
    logger.info("Printing %s", pdf_path)
    os.startfile(pdf_path, "open" if os.environ.get("IGNORE_PRINT") else "print")
