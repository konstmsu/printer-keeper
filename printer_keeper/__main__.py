import logging

import click

from printer_keeper.fortune import logger, main

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(message)s",
)


@click.command
def cli():
    try:
        main()
        logger.info("Done")
    except:
        logger.exception("Application failed")


cli()
