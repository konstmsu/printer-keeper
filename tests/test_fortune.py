import random
from printer_keeper.fortune import format_date, format_datetime, generate_fortune_html
from pathlib import Path
import re
from datetime import date, datetime, timedelta
from zoneinfo import ZoneInfo

random.seed(42)


def test_format_day():
    assert format_date(date(2022, 7, 15), include_year=False) == "пятнадцатое июля"


def test_format_datetime():
    assert (
        format_datetime(datetime(2022, 7, 15, 6, 25))
        == "пятнадцатое июля две тысячи двадцать второго года, шесть часов двадцать пять минут"
    )
    assert (
        format_datetime(datetime(2023, 1, 1, 20, 50))
        == "первое января две тысячи двадцать третьего года, двадцать часов пятьдесят минут"
    )
    assert (
        format_datetime(datetime(2023, 9, 29, 1, 31))
        == "двадцать девятое сентября две тысячи двадцать третьего года, один час тридцать одна минута"
    )
    for i in range(0, 1000):
        dt = datetime(2022, 1, 1, 13, 45) + timedelta(days=i, minutes=i)
        res = format_datetime(dt)
        assert re.search("двадцать второго|двадцать третьего|двадцать четвёртого", res)
        assert re.search(
            "января|февраля|марта|апреля|мая|июня|июля|августа|сентября|октября|ноября|декабря",
            res,
        )
        assert re.search("года", res)
        assert re.search("час|часов|часа", res)
        assert re.search("минут", res)


def test_template(snapshot):
    html = generate_fortune_html(
        "Доброе утро, мир!",
        asof=datetime(2022, 8, 4, 6, 45, tzinfo=ZoneInfo("Asia/Singapore")),
    )
    Path("generated.html").write_text(html, encoding="utf8")
    snapshot.assert_match(html)
