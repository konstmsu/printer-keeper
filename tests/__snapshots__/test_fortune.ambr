# serializer version: 1
# name: test_arithmetic_problems
  list([
    '1 + 8 = ',
    '16 + 4 = ',
    '9 + 4 = ',
    '9 + 3 = ',
    '8 - 1 = ',
    '10 - 3 = ',
  ])
# ---
# name: test_template
  '''
  <html>
    <head>
      <title>Доброе Утро!</title>
    </head>
    <body>
      <p style="font-size: 2.2em; white-space: pre-wrap"
        ><span style='color:#866103'>Ура</span>! <span style='color:#A82255'>Доброе</span> <span style='color:#5A701A'>утро</span>!
  <span style='color:#7A542E'>Сегодня</span> <span style='color:#A82255'>у</span> <span style='color:black'>нас</span> <span style='color:#77450D'>четвёртое</span> <span style='color:#5642A6'>августа</span> <span style='color:#5C255C'>2022</span>.
  
  <span style='color:#A82255'>Пословица</span> <span style='color:#7A542E'>дня</span>:
  <span style='color:#5A701A'>Кукареку</span>!
  
  <span style='color:#866103'>Чуть</span>-<span style='color:#77450D'>чуть</span> <span style='color:#5C255C'>арифметики</span>:
  □   <span style='color:#96290D'>1</span> + <span style='color:#0C5174'>1</span> =
  <span style='color:#7A542E'>Правильно</span>:    <span style='color:#7A542E'>из</span> <span style='color:blue'>1</span>
  
  <span style='color:#5642A6'>Хорошего</span> <span style='color:green'>дня</span>!</p
      >
    </body>
  </html>
  '''
# ---
